### PR TITLE
HDFS

### PR DESCRIPTION
**Pull request by mrocklin**
Starting an HDFS component to `into`.  Counterpoint to work done by @cpcloud on AWS

The goal is to 

1.  Create new types for remote data that allow us to cleanly leverage our existing logic
2.  Automate some migration and data loading on foreign hardware.

See discussion here:

https://github.com/ContinuumIO/into/issues/29
https://github.com/ContinuumIO/into/issues/34

```Python
In [1]: from into import HDFS, CSV, discover

In [2]: from pywebhdfs.webhdfs import PyWebHdfsClient

In [3]: hdfs = PyWebHdfsClient(host='123.123.123.123', port='14000', user_name='hdfs')

In [4]: data = HDFS(CSV)('/user/hive/warehouse/csv_test/data.csv', hdfs=hdfs)

In [5]: discover(data)
Out[5]: 
dshape("""var * {
  Name: string,
  RegistrationDate: datetime,
  ZipCode: int64,
  Consts: float64
  }""")
```

HDFS work done by @quasiben

### CODE REVIEW

**Comment by mrocklin** [comment]
This now also includes `SSH`

File: into/backends/ssh.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        self.auth = keyfilter(keywords(paramiko.SSHClient.connect).__contains__,
+                              kwargs)
+        self.subtype.__init__(self, *args, **kwargs)
+
+    @memoize
+    def connect(self):
+        ssh = paramiko.SSHClient()
+        ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        ssh.connect(**self.auth)
+        return ssh
```
Comment (mrocklin) [inline into/backends/ssh.py]: I'm not sure what best practices are here.  Should this close itself?  Should we not memoize?  Should we keep a connection open per-host?

File: into/backends/ssh.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+    def lines(self):
+        ssh = self.connect()
+        sftp = ssh.open_sftp()
+        return sftp.file(self.path, 'r')
+
```
Comment (mrocklin) [inline into/backends/ssh.py]: Should we have a `__str__` method here?  If so what should it look like?

File: into/backends/ssh.py (hunk 1 above)
Comment (quasiben) [inline into/backends/ssh.py]: I _assume_ we should not memoize and close the connection when done.   Not sure what happens when you leave it open -- the directive is `ssh.close`

File: into/backends/ssh.py (hunk 2 above)
Comment (quasiben) [inline into/backends/ssh.py]: suggestion for `__str__`:
<username:xxxxxxxx@hostname:port>

Comment (mrocklin) [inline into/backends/ssh.py]: This is also a file, not really an ssh connection, so we need to put path in there somehow.  Do we want to emit this string bare or wrapped with something?

Comment (quasiben) [inline into/backends/ssh.py]: Actually, I intended to write:
`<username:xxxxxxxx@hostname:port>`

Filename is a good point -- perhaps:
`<username:xxxxxxxx@hostname:port -- basename>`

**Comment by quasiben** [comment]
This looks really great!  Tests will fail when the cluster is brought down though, no?

**Comment by mrocklin** [comment]
Yes.  I have no idea how to handle testing on the HDFS code.

**Comment by mrocklin** [comment]
This is getting closer to mergability.  @cpcloud any concerns?

**Comment by mrocklin** [comment]
Passes what tests we have.

**Comment by mrocklin** [comment]
OK, merging.

**Comment by mrocklin** [comment]
@quasiben this is in.  I expect that there are lingering issues.

**Pull request merged**